### PR TITLE
introduce assert.Assert and assert.Partial()

### PR DESCRIPTION
**Pull request by i4ki**

### CODE REVIEW

File: assert/partial.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+			objtype.Name(), ofield.Name, ofield.Type,
+			targtype.Name(), tfield.Name, tfield.Type,
+			errordetails(details...),
+		)
+
+		assert.Partial(obj.Field(i).Interface(), target.Field(i).Interface())
+	}
+}
+
+func Partial(t *testing.T, obj interface{}, target interface{}, details ...interface{}) {
```
Comment (katcipis) [inline assert/partial.go:76]: details seems unused here :thinking:

File: assert/assert_test.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+			obj: struct {
+				A int
+			}{},
+			target: struct {
+				B int
+			}{},
+			fail: true,
+		},
+		{
+			name: "same struct value",
```
Comment (katcipis) [inline assert/assert_test.go:99]:
```suggestion
			name: "same struct int value",
```

File: assert/assert_test.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+		},
+		{
+			obj: struct {
+				A string
+			}{"test"},
+			target: struct {
+				A string
+			}{"test"},
+			name: "same struct field value",
+		},
```
Comment (katcipis) [inline assert/assert_test.go:125]:
```suggestion
		{
			name: "same struct string value",
			obj: struct {
				A string
			}{"test"},
			target: struct {
				A string
			}{"test"},
		},
```

This seems to be similar to the int test done before, so I imagine we are covering different types.

File: assert/assert_test.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		{
+			name: "field contains",
```
Comment (katcipis) [inline assert/assert_test.go:127]: hmm I thought that we were going just with ignoring struct fields that have 0 value on the target, but we are also doing contains on the strings ? will it also work like that for slices ? (like testing subsets ?)

I think we are missing a test for when the target has a zero value, from what we discussed before I imagine that zero values on the target are ignored.

**Review by katcipis — COMMENTED** [review]
Few comments, didn't finish the PR yet, looks cool so far :100:

File: assert/partial.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
+package assert
+
+import (
+	"reflect"
+	"testing"
+)
+
+func (assert *Assert) Partial(obj interface{}, target interface{}) {
```
Comment (katcipis) [inline assert/partial.go:8]: shouldn't we support the same formatting/description mechanism here ? or else there is no way for users to provide their own customized error message when the partial assert fails

Comment (katcipis) [reply assert/partial.go:8]: also we need to call assert.t.Helper() here no ?

File: assert/partial.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	default:
+		assert.t.Fatalf("Partial does not support comparing %s", targ.Kind())
+	}
+}
+
+func (assert *Assert) partialStruct(obj reflect.Value, target reflect.Value, details ...interface{}) {
+	objtype := obj.Type()
+	targtype := target.Type()
+
+	assert.IsTrue(targtype.Name() == objtype.Name(), "struct type mismatch.%s",
```
Comment (katcipis) [inline assert/partial.go:45]: would be good to provide each type here, or else debugging may be hard

File: assert/partial.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+		errordetails(details...))
+	assert.EqualInts(obj.NumField(), target.NumField(),
+		"number of struct fields mismatch.%s", errordetails(details...))
```
Comment (katcipis) [inline assert/partial.go:48]: provide each numbers of fields that don't match, to assist on debugging.

File: assert/partial.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+	for i := 0; i < target.NumField(); i++ {
+		ofield := objtype.Field(i)
+		tfield := targtype.Field(i)
+
+		assert.EqualBools(ofield.Anonymous, tfield.Anonymous,
+			"embedded field and non-embedded field.%s", errordetails(details...))
+
+		assert.IsTrue(ofield.Type == tfield.Type,
+			"field type mismatch: index %d (%s.%s (%s) == %s.%s (%s).%s", i,
```
Comment (katcipis) [inline assert/partial.go:58]: noice :-)

File: assert/partial.go (hunk 5 above)
Comment (katcipis) [reply assert/partial.go:8]: Just saw we pass the details on the assert constructor, but I'm a little confused, some assert methods still get details as a param for the method: https://github.com/madlambda/spells/pull/21/files?diff=unified&w=0#diff-0e9d71fd6e2e96e944d107325b0a19d470828147aa4307147e40c45c09a08344R12

It doesn't seem that tha details passed on the constructor are used: https://github.com/madlambda/spells/pull/21/files?diff=unified&w=0#diff-7695f44167b7ec89b59ff2da3c86237a9997f9974e92a6b9691e90722ed46a2cR34

And here there are no details at all as a parameter, but then the details passed on the constructor doesn't seem to be used either.

Or maybe I'm missing something 🤔

File: assert/string.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"regexp"
+	"strings"
+	"testing"
+)
+
+// StringContains asserts that string s contains the subst string and calls
+// the failure function with details otherwise.
+func (assert *Assert) StringContains(s string, substr string, details ...interface{}) {
+	assert.IsTrue(strings.Contains(s, substr), "strings.Contains(%q, %q).%s",
```
Comment (katcipis) [inline assert/string.go:12]: This is mostly cosmetic but I personally prefer this to be just like how we chain errors:

```suggestion
	assert.IsTrue(strings.Contains(s, substr), "strings.Contains(%q, %q): %s",
```

The single dot reads funny to me, but again, not necessarily bad/extremely cosmetical/feel free to disregard.

File: assert/string.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+		s, substr, errordetails(details...))
+}
+
+// StringMatch asserts that string matches the regex pattern and calls
+// the failure function with details otherwise.
+func (assert *Assert) StringMatch(pattern string, str string, details ...interface{}) {
+	matched, err := regexp.MatchString(pattern, str)
+	assert.IsFalse(err != nil, "err != nil. %v", pattern, err)
```
Comment (katcipis) [inline assert/string.go:20]: Don't we have assert.Error/assert.NoError ? they are usually quite useful when you are not interested on the specifics of errors.

**Review by katcipis — COMMENTED** [review]
Few more comments, I'm a little confused with the details being passed on the assert.New but apparently not being used, so maybe there is something missing there or I'm missing something 😆

File: assert/assert_test.go (hunk 4 above)
Comment (i4ki) [reply assert/assert_test.go:168]: yeah I added contains in the strings because this is super super useful to avoid hacks such as fixing the `wanted` strings by prepending the sandbox.RootDir of filenames in the error message.. and other hacks.. most of the time I want to check if the field contains part of the string, not it all.
I think the name makes it clear it's a Partial() match, even in the specific fields, no?

Ah yes, I did the same kind of implementation for for slices and maps, the `target` drives what's going to be asserted in the object.

Still missing a good docs.
WDYT?

File: assert/equal.go [hunk 11]
```diff
@@ -1,85 +1,135 @@
...
-		t.Fatalf("wanted[%s] but got[%s].%s", want, got, detail)
+		assert.fail("wanted[%d] but got[%d].%s", want, got, detail)
 	}
 }
 
 // EqualInts compares the two ints for equality.
-// If they are not equal t.Fatal is called using the details parameter.
-// The details parameter can be a single string of a format string + parameters.
+// If they are not equal then the Fatal() function is called with details.
 func EqualInts(t *testing.T, want int, got int, details ...interface{}) {
```
Comment (i4ki) [inline assert/equal.go:69]: I'd like to change this to int64 comparison, what do you think?

File: assert/assert_test.go (hunk 4 above)
Comment (katcipis) [reply assert/assert_test.go:168]: I think it is a good idea, was just surprised at first because I was thinking just on the zero value scenario, but like you said, in tests it is quite usual to do matching and the name of the function is already about Partial things.

Good stuff :rocket:

File: assert/equal.go (hunk 11 above)
Comment (katcipis) [reply assert/equal.go:69]: hmm won't that break a lot of code that uses just int and will have to cast it ? Usually I just use int's on code and most code I see also just uses int (the builtin `len` returns an int for example), so that could create extra work casting around if the default assert function is int64. I would be OK with providing another function for int64s, because sometimes we need those, but I think int's are more common :thinking:

**Review by katcipis — APPROVED** [review]
:rocket:

File: assert/partial.go [hunk 12]
```diff
@@ -95,48 +95,45 @@ func (assert *Assert) Partial(obj, target interface{}, details ...interface{}) {
 	}
 }
 
-func (assert *Assert) partialStruct(obj reflect.Value, target reflect.Value, details ...interface{}) {
+func (assert *Assert) partialStruct(obj interface{}, target interface{}, details ...interface{}) {
 	assert.t.Helper()
-	objtype := obj.Type()
-	targtype := target.Type()
+	elem := reflect.ValueOf(obj)
+	targ := reflect.ValueOf(target)
```
Comment (katcipis) [inline assert/partial.go:101]: This kind of stuff reminds me of allowing vars to be redeclared with another type (if done safely). I get that it can lead to confusing stuff on some scenarios, but in a lot of cases I end up on this situation where I really don't have another name for the variable (unless I go to adding the type to the name of the var :laughing: ).

**Review by katcipis — APPROVED** [review]
:rocket:

File: assert/partial.go (hunk 12 above)
Comment (i4ki) [reply assert/partial.go:101]: yeah... annoying
